woocommerce/google-listings-and-ads
# PMax Assets: Add step 2 to the campaign creation page

### Changes proposed in this Pull Request:

This PR implements parts of 📌 [Add step 2 to the campaign management page](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-campaign-step2) in #1787.

- Add a basic section for placing the assets form
- Add step 2 to the campaign creation page
- Adjust the buttons in the footer for each step
- Add `useAdaptiveFormContext` to make `AdaptiveForm`'s context also apply compatible fixes
- Add form submission related states to `AdaptiveFormContext`

### Screenshots:

https://user-images.githubusercontent.com/17420811/204020795-94cb4d0e-28f7-4434-a362-59502f3fcea0.mp4

### Detailed test instructions:

1. Go to the campaign creation page.
2. Check if the stepper works correctly.
3. Create campaigns individually using "Launch paid campaign" and "Skip adding assets" buttons.

### Additional details:

#### 💡 About the changes in `AdaptiveForm`

- I renamed `formProps` to `formContext` since it should be more precise.
- Since WC 6.9, it added the ability to obtain `Form` context via `useFormContext` hook. IMO, it's helpful to avoid prop drilling, especially since GLA codebase already has many layers of prop drilling happening on `formProps`. But this also needs 9fdd5f31983c92a211d3d640cb82c4bce2cacb6d to apply compatible fixes.
- To control the `disabled` and `loading` states of submission button, we did not have a clear way to separate form submission processing from button states. So usually, they have to be placed together (within the same component), which made data and states at different abstraction levels mixed together, and also made managing data less flexible. I added
f0563372de425278263e400e860a5fa651abce9a to explore a more proper way to use the form in WC plugin development.
   - The [processing of form submission](https://github.com/woocommerce/google-listings-and-ads/blob/28899e31daf53e79105808fc16279ed01d9b3974/js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js#L82-L115) was placed in the last step of onboarding flow rather than in the `onSubmit` callback.
   - [The `saving` state](https://github.com/woocommerce/google-listings-and-ads/blob/28899e31daf53e79105808fc16279ed01d9b3974/js/src/components/free-listings/setup-free-listings/index.js#L225) needs to be passed down to children.
   - [Mutate `formProps` to let `isSubmitting` hitchhike](https://github.com/woocommerce/google-listings-and-ads/blob/28899e31daf53e79105808fc16279ed01d9b3974/js/src/setup-ads/setup-ads-form.js#L96) to a child in several layers deep.

### Changelog entry
